File: Julia/math/magic_number.jl
"""Julia program to check if a number is a Magic Number or not
A number is said to be a magic number if the recursive sum of its digits is 1.
    For Example Let's Consider
      19, 1 + 9 = 10
      10, 1 + 0 = 1
      = 1 , Hence 19 is a Maigc Number
"""


function sum_of_digits(n)
    sum = 0
    while(n>0)
        rem = n % 10
        sum = sum + rem
        n = n ÷ 10
    end
    return sum
end

function check_magic_num(n)
    while(n > 9)
        n = sum_of_digits(n)
    end
    if(n == 1)
        return true
    else
        return false
    end
end



print("Enter the number: ")
n = readline()
n = parse(Int, n)
res = check_magic_num(abs(n))
if res
    println("The given number $n is a Magic Number.")
else
    println("The given number $n is not a Magic Number.")
end


"""
Time Complexity: O(log(n)), where 'n' is the given number
Space Complexity: O(1)

SAMPLE INPUT AND OUTPUT

SAMPLE 1
Enter the number: 23
The given number 23 is not a Magic Number.

SAMPLE 2
Enter the number: 1729
The given number 1729 is a Magic Number..

"""

File: C/math/Magic_Number.c
// A number is said to be Magic Number, when the recursive sum of the digits is 1.
#include <stdio.h>

int calsum(int);
int magic_num(int);
int main()
{
	int num;
	printf("Enter the number to be checked:");
	scanf("%d", &num);
	if (magic_num(num) == 1)
		printf("%d is a magic number", num);
	else
		printf("%d is not a magic number", num);
	return 0;
}

// Function to check if the number is a magic number or not
int magic_num(int num)
{
	int sum;
	while (num > 9)
	{
		sum = calsum(num);
		num = sum;
	}

	return num;
}

// Function to calculate the sum of digits
int calsum(int num)
{
	int sum = 0, rem;
	while (num > 0)
	{
		rem = num % 10;
		sum = sum + rem;
		num = num / 10;
	}

	return sum;
}

/*
Sample Output
Enter the number to be checked:1729
1729 is a magic number

Complexities
Time Complexity:O(n)
Space Complexity:O(n)
*/

